posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,340 @@
+                ],
+            )
+            fields.append(
+                ast.Alias(alias="breakdown_values", expr=ast.Array(exprs=[datediff_call])),
+            )
+
+        event_filters = [
+            self.entity_to_expr(
+                entity=self.target_entity
+                if event_query_type == RetentionQueryType.TARGET
+                or event_query_type == RetentionQueryType.TARGET_FIRST_TIME
+                else self.returning_entity
+            )
+        ]
+
+        if self.query.properties is not None and self.query.properties != []:
+            event_filters.append(property_to_expr(self.query.properties, self.team))
+
+        if (
+            self.query.filterTestAccounts
+            and isinstance(self.team.test_account_filters, list)
+            and len(self.team.test_account_filters) > 0
+        ):
+            for prop in self.team.test_account_filters:
+                event_filters.append(property_to_expr(prop, self.team))
+
+        date_filter_expr = self.date_filter_expr(event_query_type)
+        if event_query_type != RetentionQueryType.TARGET_FIRST_TIME:
+            event_filters.append(date_filter_expr)
+
+        group_by_fields = None
+        having_expr = None
+        if event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            group_by_fields = [ast.Field(chain=["target"])]
+            having_expr = date_filter_expr
+        elif event_query_type == RetentionQueryType.RETURNING:
+            group_by_fields = [ast.Field(chain=["target"]), ast.Field(chain=["event_date"])]
+
+        result = parse_select("SELECT * FROM events", timings=self.timings)
+        result.select = fields
+        result.distinct = event_query_type == RetentionQueryType.TARGET
+        result.where = ast.And(exprs=event_filters)
+        result.group_by = group_by_fields
+        result.having = having_expr
+
+        if self.query.samplingFactor is not None and isinstance(self.query.samplingFactor, float):
+            result.select_from.sample = ast.SampleExpr(
+                sample_value=ast.RatioExpr(left=ast.Constant(value=self.query.samplingFactor))
+            )
+
+        return result
+
+    def entity_to_expr(self, entity: dict) -> ast.Expr:
+        if entity["type"] == TREND_FILTER_TYPE_ACTIONS and entity["id"] is not None:
+            action = Action.objects.get(pk=entity["id"])
+            return action_to_expr(action)
+        elif entity["type"] == TREND_FILTER_TYPE_EVENTS:
+            if entity["id"] is None:
+                return ast.Constant(value=True)
+
+            return ast.CompareOperation(
+                op=ast.CompareOperationOp.Eq,
+                left=ast.Field(chain=["events", "event"]),
+                right=ast.Constant(value=entity["id"]),
+            )
+
+        return ast.CompareOperation(
+            op=ast.CompareOperationOp.Eq,
+            left=ast.Field(chain=["events", "event"]),
+            right=ast.Constant(value=PAGEVIEW_EVENT),
+        )
```

Comment:
I think this could be some utility function in hogql/property.py, as I think both trends and lifecycle have somewhat identical code somewhere.

Reply:
I didn't find code that does it like this, or are you suggesting to move it there?

Reply:
This seems like something that'll be common to all insights, as most offer you an "entity" (action vs event) to choose from. It's not yet abstracted.